packages/block-library/src/navigation-link/hooks.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import {
+	category,
+	page,
+	postTitle,
+	tag,
+	customPostType,
+} from '@wordpress/icons';
+
+/**
+ * Internal dependencies
+ */
+import fallbackVariations from './fallback-variations';
+
+function getIcon( variationName ) {
+	switch ( variationName ) {
+		case 'post':
+			return postTitle;
+		case 'page':
+			return page;
+		case 'tag':
+			return tag;
+		case 'category':
+			return category;
+		default:
+			return customPostType;
+	}
+}
+
+function enhanceNavigationLinkVariations( settings, name ) {
+	if ( name !== 'core/navigation-link' ) {
+		return settings;
+	}
+
+	// Fallback handling may be deleted after supported WP ranges understand the `variations`
+	// property when passed to register_block_type_from_metadata in index.php
+	if ( ! settings.variations ) {
+		return {
+			...settings,
+			variations: [ ...fallbackVariations ],
+		};
+	}
+
+	// Otherwise decorate server passed variations with an icon and isActive function
+	if ( settings.variations ) {
+		const variations = settings.variations.map( ( variation ) => {
+			return {
+				...variation,
+				...( ! variation.icon && {
+					icon: getIcon( variation.name ),
+				} ),
+				...( ! variation.isActive && {
+					isActive: ( blockAttributes, variationAttributes ) => {
+						return (
+							blockAttributes.type === variationAttributes.type
+						);
+					},
+				} ),
+			};
+		} );
```

Comment:
_Nitpick_ — I often have a hard time reading spreads, particularly if they're nested; how would you feel about something like the following?

```suggestion
		const variations = settings.variations.map( ( variation ) => {
			const isActive = ( blockAttributes, variationAttributes ) =>
				blockAttributes.type === variationAttributes.type;

			return Object.assign(
				variation,
				! variation.icon && { icon: getIcon( variation.name ) },
				! variation.isActive && { isActive }
			);
		} );
```

I think moving the function definition out of the final object is also good for readability.

Reply:
Thanks @vcanales I try to default to project convention for readability. In this case there's mixed usage, so I'll keep the spread props since it looks like we want to return a new object.

The cleanest usage is probably using lodash assigns like in the [default example](https://developer.wordpress.org/block-editor/developers/filters/block-filters/#blocks-registerblocktype), but I wanted to avoid pulling that in as a dep.